src/pybind/mgr/orchestrator/_interface.py

```diff
@@ -1092,6 +1126,29 @@ def to_json(self) -> OrderedDict:
             out['events'] = [e.to_json() for e in self.events]
         return out
 
+    def to_dict(self) -> OrderedDict:
+        out = self.spec.to_json()
+        status = {
+            'container_image_id': self.container_image_id,
+            'container_image_name': self.container_image_name,
+            'rados_config_location': self.rados_config_location,
+            'service_url': self.service_url,
+            'size': self.size,
+            'running': self.running,
+            'last_refresh': self.last_refresh,
+            'created': self.created,
+            'virtual_ip': self.virtual_ip,
+            'ports': self.ports if self.ports else None,
+        }
+        for k in ['last_refresh', 'created']:
+            if getattr(self, k):
+                status[k] = datetime_to_str(getattr(self, k))
+        status = {k: v for (k, v) in status.items() if v is not None}
```

Comment:
Maybe it's just me but I prefer to have clear declarative data struct definitions rather than ones that are dynamically stretched:
```suggestion
        status = {
            'container_image_id': self.container_image_id,
            'container_image_name': self.container_image_name,
            'rados_config_location': self.rados_config_location,
            'service_url': self.service_url,
            'size': self.size,
            'running': self.running,
            'last_refresh': datetime_to_str(self.last_refresh),
            'created': datetime_to_str(self.created),
            'virtual_ip': self.virtual_ip,
            'ports': self.ports if self.ports else None,
        }
        # REMOVE the following?
        # status = {k: v for (k, v) in status.items() if v is not None}
```
Why do we remove the keys that are None? For data sharing it's better to have stable/fixed data structure that ones that randomly show some properties...

@sebastian-philipp @jmolmo what do you think?

Reply:
Removed this for now.

Reply:
Same here. This is new code. We should deliver clean new code, and not follow suboptimal code.

Reply:
@epuertat your suggestion is also [favoring explicit over implicit](https://www.python.org/dev/peps/pep-0020/#id2), which is something we should follow :)